lkcrypto.c

```diff
@@ -213,6 +262,85 @@ static int decrypt_encrypt(lua_State *L, int dec_enc)
 	}
 }
 
+static int decrypt_encrypt_aead(lua_State *L, int dec_enc)
+{
+	size_t key_len = 0;
+	size_t iv_len = 0;
+	size_t auth_len = 0;
+	size_t ciphertext_len = 0;
+	size_t aad_len = 0;
+
+	int result = 0;
+	int nargs = 0;
+	const u8 *key;
+	const u8 *iv_orig;
+	const u8 *auth_orig;
+	const u8 *aad_orig;
+	const u8 *ciphertext;
+	u8 iv[AES_MAX_KEY_SIZE] = {
+		0,
+	};
+	u8 auth[MAX_AUTH_SIZE] = {
+		0,
+	};
+	u8 *buffer = NULL;
+	u8 *aad = NULL;
+
+	struct cipher_aead_tfm_t *tfm = NULL;
+
+	nargs = lua_gettop(L);
+
+	tfm = luaL_checkudata(L, 1, CUJO_CIPHER_AEAD_TFM_LUA_NAME);
+	key = luaL_checklstring(L, 2, &key_len);
+	ciphertext = luaL_checklstring(L, 3, &ciphertext_len);
+	iv_orig = luaL_checklstring(L, 4, &iv_len);
+	memcpy(iv, iv_orig, iv_len);
+	if (dec_enc == 0) {
+		auth_orig = luaL_checklstring(L, 5, &auth_len);
+		memcpy(auth, auth_orig, auth_len);
+	}
+
+	buffer = kmalloc(ciphertext_len, GFP_ATOMIC);
+
+	if (dec_enc == 1) {
```

Comment:
This seems like it'd be cleaner as an `else` above, keeping the argument handling together.